include/api/CDataFrameAnalysisRunner.h

```diff
@@ -67,9 +72,25 @@ class API_EXPORT CDataFrameAnalysisRunner {
     //! into main memory during an analysis.
     virtual std::size_t numberOfPartitions() const = 0;
 
-    //! \return The number of columns this analysis requires. This includes
-    //! the columns of the input frame plus any that the analysis will append.
-    virtual std::size_t requiredFrameColumns() const = 0;
+    //! \return The number of columns this analysis appends.
+    virtual std::size_t numberExtraColumns() const = 0;
+
+    //! Write the extra columns of \p row added by the analysis to \p writer.
+    //!
+    //! This should create a new object of the form:
+    //! <pre>
+    //! {
+    //!   "name of column n":   "value of column n",
+    //!   "name of column n+1": "value of column n+1",
+    //!   ...
+    //! }
+    //! </pre>
+    //! with one named member for each column added.
+    //!
+    //! \param[in] row The row to write the columns added by this analysis.
+    //! \param[in,out] writer The stream to which to write the extra columns.
+    virtual void writeOneRow(TRowRef row,
+                             core::CRapidJsonConcurrentLineWriter& writer) const = 0;
```

Comment:
can the writer change per call? if not, can it be set at construction?

Reply:
This is the object created [here](https://github.com/elastic/ml-cpp/pull/334/files#diff-740e94a98f6c2a36c617970df2bc2c81R213) so not currently available at construction time of `CDataFrameAnalysisRunner`. We could refactor so it is a member of `CDataFrameAnalyzer` and only created once (rather than once per analysis, although currently this is equivalent). I think this would be a good idea, but I don't want to make that refactor now.

That aside I'm not sure I'd want to make this a member of the `CDataFrameAnalysisRunner`. Really, this function shouldn't require state from any implementation object. It isn't static in order to be virtual, because different analyses will want to extract different column values. However, all the information is coming from the row passed in and some local statics. So this feels extrinsic to a particular instance of the class and being forced to always create a row writer whenever you create an object feels undesirable to me.

Reply:
understood, I am asking because this looks like the other extreme: supplying the write for _every_ row. Looks like the reason for this is the extra wrapping required in `writeResultsOf`.

What about just accessing the data and doing the json'ification in `writeResultsOf` , which would also get rid of forward declarations.

Just suggestions.

Reply:
The problem is we'd then need logic in this function for every single type of analysis. I quite like that the CDataFrameAnalyzer class is agnostic to the type of analyses it is running. To keep this it has to somehow pass off writing analysis specific parts of the results to the `CDataFrameAnalysisRunner`.

If I understand your suggestion correctly, we'd have to have a something like a big switch statement in writeResultsOf and some sort of  identifier returned by each analyser.

Equally if all the logic in `writeResultsOf` gets pushed down into `CDataFrameAnalysisRunner` implementations we'd duplicate it for each different analysis.

Reply:
> If I understand your suggestion correctly, we'd have to have a something like a big switch statement in writeResultsOf and some sort of identifier returned by each analyser.

sorry, I do not get that. All I see at the moment is writing 1 key and 1 value. No need for a big switch statement.

Reply:
I mean down the line when we have additional analysis types (regression, etc) and these will all have their own runners (say CDataFrameXGBoostRunner, etc). Each type of analysis will have its own column names and numbers of columns they want to write. This information can all be in the `writeResultsOf`, but not without doing something like a switch statement. Additional analysis types are definitely coming, so this isn't premature generalisation.